grakn-graql/src/test/java/ai/grakn/graql/internal/query/QueryPlannerTest.java

```diff
@@ -91,6 +116,172 @@ public void setUp() {
+
+        pattern = and(
+                x.isa(thingy),
+                y.isa(thingy2),
+                z.isa(thingy4),
+                var().rel(x).rel(y).rel(z));
+        plan = getPlan(pattern);
+        assertEquals(4L, plan.stream().filter(LabelFragment.class::isInstance).count());
+        assertEquals(4L, plan.stream()
+                .filter(fragment -> fragment instanceof OutIsaFragment || fragment instanceof InIsaFragment).count());
+    }
+
+    @Test
+    public void inferRelationshipTypeWhereAVarHasTwoTypes() {
+        Pattern pattern;
+        ImmutableList<Fragment> plan;
+
+        pattern = and(
+                x.isa(thingy),
+                x.isa(thingy1),
+                y.isa(thingy4),
+                var().rel(x).rel(y));
+
+        plan = getPlan(pattern);
+        assertEquals(4L, plan.stream().filter(LabelFragment.class::isInstance).count());
+        assertEquals(4L, plan.stream()
+                .filter(fragment -> fragment instanceof OutIsaFragment || fragment instanceof InIsaFragment).count());
+        String relationship = plan.get(4).start().getValue();
+        assertNotEquals(relationship, x.getValue());
+        assertNotEquals(relationship, y.getValue());
+    }
+
+    @Test
+    public void inferRelationshipTypeWhereAVarHasIncorrectTypes() {
+        Pattern pattern;
+        ImmutableList<Fragment> plan;
+
+        pattern = and(
+                x.isa(veryRelated),
+                x.isa(thingy2),
+                y.isa(thingy4),
+                var().rel(x).rel(y));
+        plan = getPlan(pattern);
+        assertEquals(3L, plan.stream().filter(LabelFragment.class::isInstance).count());
+        assertEquals(3L, plan.stream()
+                .filter(fragment -> fragment instanceof OutIsaFragment || fragment instanceof InIsaFragment).count());
+    }
+
+    @Test
+    public void avoidImplicitTypes() {
+        Pattern pattern;
+        ImmutableList<Fragment> plan;
+
+        pattern = and(
+                x.isa(thingy2),
+                y.isa(thingy4),
+                var().rel(x).rel(y));
+        plan = getPlan(pattern);
+        assertEquals(3L, plan.stream().filter(LabelFragment.class::isInstance).count());
+        String relationship = plan.get(4).start().getValue();
+
+        // should  start from relationship
+        assertNotEquals(relationship, x.getValue());
+        assertNotEquals(relationship, y.getValue());
+
+        pattern = and(
+                x.isa(resourceType),
+                y.isa(thingy4),
+                var().rel(x).rel(y));
+        plan = getPlan(pattern);
+        assertEquals(3L, plan.stream().filter(LabelFragment.class::isInstance).count());
+        relationship = plan.get(4).start().getValue();
+
+        // should start from role player
+        assertTrue(relationship.equals(x.getValue()) || relationship.equals(y.getValue()));
+        assertTrue(plan.get(5) instanceof OutIsaFragment);
+    }
+
```

Comment:
I have a potentially useful test set defined in `typeInferenceTest.gql` and unit tests in `TypeInferenceQueryTest`. You might want to have a look.

Reply:
👍